hive-bigquery-connector-common/src/main/java/com/google/cloud/hive/bigquery/connector/utils/hive/HiveUtils.java

```diff
@@ -45,9 +44,16 @@ public static boolean isExternalTable(Table table) {
     return "TRUE".equalsIgnoreCase(params.get("EXTERNAL"));
   }
 
-  /** Returns the ID of the Hive query as set by Hive in the configuration. */
   public static String getQueryId(Configuration conf) {
-    return HiveConf.getVar(conf, HiveConf.ConfVars.HIVEQUERYID);
+    String hiveQueryId = HiveConf.getVar(conf, HiveConf.ConfVars.HIVEQUERYID, "");
```

Comment:
maybe change this to getWorkflowId, both hive and pig set property "mapreduce.workflow.id", hive set as hive_{query_id}, pig set to pig_{pig_id}, so here can just get from this property.

Reply:
I think it's worth checking each conf property (`hive.query.id` and `pig.script.id`) instead of `mapreduce.workflow.id`. That way we can set a more specific prefix to the returned query ID.

That said, I'm not against renaming the function, as "workflow" is more generic than "query" and "script".

Reply:
@yigress BTW, should we then rename `JobUtils.getQuerySubDir()` to `getWorkflowSubDir()`?

Reply:
my concern is pig.script.id is not necessarily set in some scenarios, though i don't know if by the time when hive-bq-connector checks it it is always set or not, but the mapreduce.workflow.id seems to be set. since hive and pig they set with prefix hive_query_id and pig_id, it may be better we just take it to be consistent with them?

Reply:
From what I can see, the `hive.query.id` and `pig.script.id` are of the form respectively `<username>_<timestamp>_<random_string>` and `pig_<random_string>`. So from the values themselves you can't really tell explicitly that it might be coming from a Hive query. Also, if we can later add support for Spark SQL, then we'll have the same issue where we have to take the ID from `spark.app.id` and potentially some other things like stage ID.

Reply:
In what scenarios do you see the Pig script ID not being set? From what I could tell, one case is when we instantiate the output committer, but that doesn't have an influence as the query ID isn't needed by the connector in that case.

Also, at the end of `getQueryId()`, we have this line:

```
throw new RuntimeException("No query id found in Hadoop conf");
```

So that protects us from cases where the values aren't set as expected. The job would fail loudly. That failure doesn't occur in our integration tests, so that's why I don't think we need to worry about this.

Reply:
hive have conf.set("mapreduce.workflow.id", "hive_" + queryId); and pig have conf.set("mapreduce.workflow.id", "pig_" + id); so it is able to differentiate pig or hive, but that being side, this is a mr property, so I am fine with current logic.

Reply:
Oh I see, you're correct. I just verified that in the integration tests as well. I think it's fine to make that change then. I'll submit a change to the PR.

Reply:
Ok I've pushed the change. It will be interesting to see how that affects things with Spark SQL. Maybe that'll make things easier.

Reply:
So it looks like after this change the integration tests are failing. This is due to the fact that, somehow, the workflow ID doesn't remain consistent over time:

In `BigQueryStorageHandlerBase.configureOutputJobProperties()` we have:
```
hive.query.id: jphalip_20240227101551_29776b3a-9424-4808-9064-830f219a5940
mapreduce.workflow.id: hive_jphalip_20240227101548_9bc52357-a668-47b2-aeba-d20b38eaa288
```

Then soon after in `BigQueryMetaHookBase.preInsertTable()`:

```
hive.query.id: jphalip_20240227101551_29776b3a-9424-4808-9064-830f219a5940
mapreduce.workflow.id: hive_jphalip_20240227101551_29776b3a-9424-4808-9064-830f219a5940
```

I'm not sure why in `BigQueryStorageHandlerBase.configureOutputJobProperties()` the workflow ID would be different and not include the actual query ID. @yigress any thoughts?

Reply:
not sure of the root cause. in that case, maybe we just recovert back to previous logic, and add extra check of workflow id when both hive and pig checks fail?

Reply:
It appears that the `mapreduce.workflow.id` property is set in `org.apache.hadoop.hive.ql.Driver.compile()`. However, it also appears that that method is executed **after** the storage handler's `configureOutputJobProperties()`. So the storage handler ends up seeing the `mapreduce.workflow.id` value that was set for the previously executed query (the `CREATE TABLE` statement) instead of the currently executed query (the `INSERT` statement). The `hive.query.id` property, however is consistent throughout.

So unfortunately it looks like the storage handler cannot rely on the `mapreduce.workflow.id` property when creating the JobDetails JSON file.

Regarding what you said: "add extra check of workflow id when both hive and pig checks fail". Could you clarify what you mean? In what case would that be useful/necessary?

Reply:
@yigress I've reverted the change and now the tests are passing. Does it look ok to you, or do you still want to add those extra checks, and if so, could you clarify?

Reply:
i just noticed that pig actually shares one pig id in a pig sesison, a pig session could have multiple jobs, kind of similar to spark. so using pig id would have issues

Reply:
@yigress Good catch. I just changed it so that, when using Pig, we actually use the mapreduce workflow ID as Pig uses it consistently through out the execution of the job. I've added some comments to the HiveUtils function as well, and added some unit tests. Let me know if that looks good to you.

My hope is that we'll be able to do something similar for Spark. We can look into that later.

Reply:
are you able to confirm the workflow id is different for different jobs inside same pig session/id?

Reply:
@yigress Hmm, so the mapreduce workflow ID actually just is "pig_" followed by the pig script id, so actually not that useful. However, I see that there is the `pig.job.submitted.timestamp` property. This is set here: https://github.com/apache/pig/blob/3e896e5d588273f889640b535d2daa48c5e079f8/src/org/apache/pig/backend/hadoop/executionengine/mapReduceLayer/MapReduceLauncher.java#L289-L290

So we could use that, assume that 2 jobs cannot possibly have the same timestamp value. I was able to verify that in the integration tests at least.

Reply:
sg

Reply:
@yigress To clarify, I propose we use return something like: `String.format("pig-%s-%s", conf.get("pig.script.id"), conf.get("pig.job.submitted.timestamp"))`

Reply:
@yigress Ok just made the change